### ISSUE TITLE
Prime meridian information seems to be ignored in some cases

### ISSUE THREAD

**Issue by calvinmetcalf** [proj4js/proj4js#259]
Hi,

I have been trying to process data in shapefile format with the attached .prj file in well-known text format (wkt).

test_case_prj.txt

The results I get are wrong: they are shifted to the West. With exactly the same input data, QGIS displays data correctly.

I soon became aware that 'long0' was not initialized, though it is used in the forward and inverse functions for Lambert Conformal Conic transformations.

If I set it to the value held in the prj file converted in radians in the initialization function associated to 'lcc':

          if (! this.long0) {
            this.long0 = 0.04079234433; // This is 2° 20' 14,025" East
          }

then I get the expected results.

I also noticed that the following lines were commented in function cleanWKT:

if (wkt.type === 'GEOGCS') {
	    geogcs = wkt;
	  }
	  if (geogcs) {
	    //if(wkt.GEOGCS.PRIMEM&&wkt.GEOGCS.PRIMEM.convert){
	    //  wkt.from_greenwich=wkt.GEOGCS.PRIMEM.convert*D2R;
	    //}
	    if (geogcs.DATUM) {

If I uncomment the three lines above instead of forcing the value of 'long0', then I also get the expected results.

It looks like the prime meridian information held in the prj file is simply ignored:

PROJCS["Lambert_Conformal_Conic",
GEOGCS["GCS_unnamed ellipse",
DATUM["D_unknown",SPHEROID["Unknown",6378249.2,293.466020999993]],
PRIMEM["Unknown",2.337229167],
UNIT["Degree",0.017453292519943295]],

Any help or insight would be highly appreciated.

Many thanks and kind regards.

moved from calvinmetcalf/shapefile-js#81

**Comment by calvinmetcalf** [comment]
ok @starsway next step will be for me to figure out why that line was commented out

**Comment by starsway** [comment]
Again, thank you very much.
I attempted to answer this question and I found out that this line was already commented the first time it appeared in the code in the following commit:

https://github.com/proj4js/proj4js/commit/16328f8d0150d59e1473698310c4a1706833aae4

**Comment by starsway** [comment]
Please ignore my previous comment. The commented line is older than what I wrote. It dates back at least to:

https://github.com/proj4js/proj4js/commit/f38b18f6355dc9de5d2b43989946c2aa54568e86

**Comment by calvinmetcalf** [comment]
so just uncommenting (and fixing) causes this projection to be shifted about 17 degrees

> PROJCS["S-JTSK (Ferro) / Krovak", GEOGCS["S-JTSK (Ferro)", DATUM["D_S_JTSK", SPHEROID["Bessel_1841",6377397.155,299.1528128]], PRIMEM["Ferro",-17.66666666666667], UNIT["Degree",0.017453292519943295]], PROJECTION["Krovak"], PARAMETER["latitude_of_center",49.5], PARAMETER["longitude_of_center",42.5], PARAMETER["azimuth",30.28813972222222], PARAMETER["pseudo_standard_parallel_1",78.5], PARAMETER["scale_factor",0.9999], PARAMETER["false_easting",0], PARAMETER["false_northing",0], UNIT["Meter",1]]

checking against cs2cs our original result is correct so this introduces an error so looks like we'd need a more nuanced fix.  It would be helpful if you could find any documented projections that have this issue that would also be helpful

**Comment by starsway** [comment]
Thank you very much for digging into this and for your answer.

Sure, I would be glad to help in any way I can.

I just need to make sure I understand correctly current situation and the implications of your observation.

- I am under the impression that the 'PRIMEM' is read but never used. Am I missing something or is this correct? In case it is correct, then this does not sound quite right. I would expect the prime meridian parameter to be used in some way... But I may be wrong and I may have totally overlooked something.
- The projection you have mentioned does indeed have a prime meridian with longitude -17.6666666666 according to https://epsg.io/4818 ; Czech Republic lies between 12°E and 19°E, so it is easy to check whether locations supposed to be there remain there. That's a reassuring point to have while investigating this. Now, could the shift you have observed be caused by the fact that there is already a custom processing for S-JTSK (Ferro) triggered by this line:

	exports$1.ferro = -17.666666666667; //"17d40'W",

leading points to be shifted twice?

**Comment by calvinmetcalf** [comment]
yeah so the issue is that we're currently picking up the named prime meridians, and not looking at their actual value, the reason you projection isn't working is you have 'Unknown' instead of Paris for the name of your prime meridian, just commentating those lines doesn't work because it causes both the named version and the value to be used, we'd probably want it so it only uses that value if it doesn't have the one in it's list of constants

**Comment by calvinmetcalf** [comment]
or maybe not the case, still investigating

**Comment by starsway** [comment]
Thanks. I'm looking at this piece of code inside function 'parseProj':

```
	  for (paramName in paramObj) {
	    paramVal = paramObj[paramName];
	    if (paramName in params) {
	      paramOutname = params[paramName];
	      if (typeof paramOutname === 'function') {
	        paramOutname(paramVal);
	      }
	      else {
	        self[paramOutname] = paramVal;
	      }
	    }
	    else {
	      self[paramName] = paramVal;
	    }
	  }
	  if(typeof self.datumCode === 'string' && self.datumCode !== "WGS84"){
	    self.datumCode = self.datumCode.toLowerCase();
	  }
	  return self;

```

I was wondering whether it would make any sense to check just before returning whether 'long0' and 'from_greenwich' are both undefined. In such a case, we would check for the PRIMEM parameter and set one of them to its value when it is present.

**Comment by starsway** [comment]
Wrong lead - sorry - please ignore my previous suggestion.

**Comment by starsway** [comment]
I first thought that inserting a slightly modified version of the currently commented lines would do the trick:

```
	    if(wkt.GEOGCS.PRIMEM&&wkt.GEOGCS.PRIMEM.convert&&(wkt.from_greenwich === undefined)){
	      wkt.from_greenwich=wkt.GEOGCS.PRIMEM.convert*D2R;
	    }
```

but then I discovered that 'long0' was hard-coded in [krovak.js](https://github.com/proj4js/proj4js/blob/v2.0.1/lib/projections/krovak.js).

```
  if (!this.long0) {
    this.long0 = 0.7417649320975901 - 0.308341501185665;
  }

```
(- 0.308341501185665 rad = -17 - 2 / 3 deg) So I guess all these hard-coded prime meridian definitions should be removed before starting to read the prime meridian information PRIMEM found in the prj files...

**Comment by calvinmetcalf** [comment]
well except I don't know if some krovak omit the primem is the problem (or
give an incorect one) I'd be interested to know where you got that
shapefile from and/or how that proj file was generated

**Comment by starsway** [comment]
This shapefile was handed to me by a prospective vendor and I have no (and cannot have any) additional information about how it was generated, unfortunately... And QGIS is processing it seamlessly.

For the time being, I have patched my code so that data get correctly displayed for the small set of shapefiles I have to tackle, so I do not really need anything more. I'm lucky since I have only a very limited number of files to display, so I don't need to find a general situation. I understand that, of course, things are much more difficult if we are to find a general, genuine patch to proj4js code that does not break you current test suite.

However, it is my opinion that the fact that PRIMEM is **always** ignored in proj4js is not completely satisfying. I understand you want to address a situation where 'some krovak omits the primem'. Maybe we should then at least verify whether information about 'central_meridian' or PRIMEM has been provided before forcing their value in the code... I admit this is not an easy task.

Anyway, I appreciate your help. Thanks very much.

**Comment by calvinmetcalf** [comment]
I'm 100% with you that there is a bug here in the parsing, it's just an old project so we need to make sure that we don't introduce any new bugs when we fix old ones as there are just so many layers where things happen, in this case the parser which is parsing the proj string, the transformer, which is where the from_greenwich value is handled and the specific projection, in this case lcc or krovak and the layers are handling the values in different ways so the fix is likely to be more complex then it would seem at first glance
